feat(Pills): add Pills spec (#17017)
microsoft/fluentui: specs/Pills.md

# [Pills]()

## Background

> PIlls should be used when representing an input, as a way to filter content, or to represent an attribute.

Reviewer: nit
```suggestion
Pills should be used when representing an input, as a way to filter content, or to represent an attribute.
```

## Prior Art

- [Evergreen](https://evergreen.segment.com/components/badge-and-pill/)
- [Lightning](https://www.lightningdesignsystem.com/components/pills/)
- [Material UI](https://material-ui.com/components/chips/#chip)
- [OpenUI Research](https://github.com/WICG/open-ui/pull/259)

## Sample Code

```jsx

> <Pill>Item 1</Pill>

Reviewer 2: aren't we missing `<Pills>` in the example? The goal of this work should be to create both container and item components.

Author: Added

<Pill clickable>Item 2</Pill>

> <TogglePill>On</Pill>

Reviewer 3: nit
```suggestion
<TogglePill>On</TogglePill>
```

- Appearance: `Filled Grey (filled)`, `Filled White (inverted)`, `Outline`

## PROPS

```typescript
export type PillShape = 'rounded' | 'circular';

export type PillAppearence = 'filled' | 'inverted' | 'outline';

export type SizeValue = 'smaller' | 'small' | 'medium';

export interface PillProps extends ComponentProps, React.HTMLAttributes<HTMLElement> {
  /**
   * A Pill can be sized.
   */
  size?: SizeValue;

  /**
   * A Pill can be circular or rounded
   */
  shape?: PillShape;

  /**
   * A Pill can be filled, inverted or outline
   */
  appearance?: PillVariant;

  /**
   * Media slot
   */
  media?: ShorthandProps<HTMLElement>;

  /**
   * Details that will appear as second line
   */
  details?: ShorthandProps<HTMLElement>;

  /** A Pill can be dismissible. */
  dismissible?: boolean;

  /** A Pill can be disabled. */
  disabled?: boolean;

  /** A Pill can be selected. */
  selected?: boolean;

  /** A Pill can be clickable */
  clickable?: boolean;

  /** A callback to be called when Pill is clicked */
  onClick?: function;

  /**
   * A button shorthand for the dismiss action slot. To use this slot the pill should be
   * dismissible.
   */

>   dismissIcon?: ShorthandProps<HTMLElement>;

Reviewer: Is `dismissIcon` mandatory for `<Pill dismissible />? If not, is there a default dismiss icon or no icon is rendered?

Author: By default no icon is rendered, if the `Pill` is dismissible we render the default `X` icon or what users provides.

}

export type TogglePillProps = Omit<PillProps, 'clickable'>;
```

## Structure

- _**Public**_

```jsx
<Pill />

<Pill clickable />

<TogglePill />
```

- _**DOM**_

```html
<span>
  ...
</span>

<span role="button">
  ...
</span>

<span role="switch" aria-checked>
  ...
</span>
```

## Migration

- _Migration from v8_

There's no components related to `Pill` in V8

- _Migration from v0_

> `Dropdown` component should use `Pills` over `DropdownSelectedItem`.

Reviewer 2: maybe, but I would make that as a separate work item to migrate.

Author: Do you mean to create a separated document for the migration ( similar to what Ben did for Avatar ) or only talking about the working item itself?

Author: Should I remove it from the spec?

### Pill with dismiss

#### Accessibility

> - Tab key to set focus on the Pill

Reviewer 2: during the design review, listbox/option were proposed - please check also the pills in dropdown. With that, arrow key navigation would be used to navigate. We should even do left+up arrow to move to previous, right+down arrow to move to next.

Author: I also had conversation with @kolaps33 about it, can you express your opinions  again @kolaps33

Reviewer 2: we should not rely on using plain divs to form appropriate strucutre. We need to have `Pills` component that defines keyboard navigation and roles properly. Modifications can be done using behaviors, not primarily using role overrides.

In most cases, we want to reduce the amount of tab stops - that is typically achieved by using arrow keys in places where possible.

Author: Updated to: 

`- Arrow keys navigates between `Pill``

- Tab key moves focus to next `Pill`
- Esc key dismiss the `Pill`

#### Keyboard Navigation

| key                | state    | title                                                              |
| ------------------ | -------- | ------------------------------------------------------------------ |

> | left/right         | OnFocus  | Role= “button”, aria-describedby=”Press enter or delete to remove” |

Reviewer 2: also up/down can work. For roles, listbox/option were proposed. The instruction message needs to come from a prop to allow translation

Reviewer 2: on the line 130, tabbing is proposed to navigate among pill, or is that to navigate among pill containers?

Author: It should be to navigate among components, although I recall seeing in the design spec telling about navigating between pills

Author: The design spec actually says that should navigate between items

Reviewer 2: During review, listbox/options roles were proposed which are connected with arrow key navigation. If this has not been reflected in the design, we need to ask designers to apply this change.

> | Enter/Delete       | OnDelete | The pill is removed since there is only one primary action         |

Reviewer 2: +Space

> | Enter/Delete/Space | OnDelete | The pill is removed since there is only one primary action         |

Reviewer: Is it really a common practice for Enter/Space to dismiss the pill? Especially when you can have a pill with popup where Enter/Space opens the popup.

Author: I don't think it actually it's but this is specified in the design spec

Reviewer: cc @jurokapsiar @kolaps33

> | Backspace/Ctrl + X | OnDelete | Backspace/Ctrl + X                                                 |

Reviewer 2: deleting should work only if it is dismissible pill

### Pill with popup

#### Accessibility

- Popup opens after Xms of mouse hover state
- Clicking in the Pill open popup and clicking outside should dimiss it

#### Keyboard Navigation

| key                | state    | title                                                        |
| ------------------ | -------- | ------------------------------------------------------------ |
| Esc                | OnFocus  | Dismis popup                                                 |

> | Enter              | OnFocus  | Open Popup                                                   |

Reviewer 2: +Space (for other variants as well)

# [Pills]()

## Background

PIlls should be used when representing an input, as a way to filter content, or to represent an attribute.

## Prior Art

- [Evergreen](https://evergreen.segment.com/components/badge-and-pill/)
- [Lightning](https://www.lightningdesignsystem.com/components/pills/)

> - [OpenUI Research]()

Author: ```suggestion
- [OpenUI Research](https://github.com/WICG/open-ui/pull/259)
```

</Pills>
```

## Variants

- Shape: `circular`, `rounded`
- Size: `smaller`, `small`, `medium`
- Appearance: `Filled Grey (filled)`, `Filled White (inverted)`, `Outline`

## PROPS

```typescript
export type PillShape = 'rounded' | 'circular';

export type PillAppearence = 'filled' | 'inverted' | 'outline';

export type SizeValue = 'smaller' | 'small' | 'medium';

export interface PillProps extends ComponentProps, React.HTMLAttributes<HTMLElement> {
  /**
   * A Pill can be sized.
   */
  size?: SizeValue;

  /**
   * A Pill can be circular or rounded
   */
  shape?: PillShape;

  /**
   * A Pill can be filled, inverted or outline
   */
  appearance?: PillVariant;

  /**
   * Media slot
   */
  media?: ShorthandProps<HTMLElement>;

  /**
   * Details that will appear as second line
   */
  details?: ShorthandProps<HTMLElement>;

  /** A Pill can be dismissible. */
  dismissible?: boolean;

  /** A Pill can be disabled. */
  disabled?: boolean;

  /** A Pill can be selected. */
  selected?: boolean;

  /** A Pill can be draggable  */
  draggable?: boolean;

  /**
   * A button shorthand for the dismiss action slot. To use this slot the pill should be
   * dismissible.
   */

>   dismissAction?: ShorthandObject;

Author: ```suggestion
  dismissAction?: ShorthandProps<HTMLElement>;
```

>   dismissAction?: ShorthandValue<HTMLElement>;

Author: ```suggestion
  dismissAction?: ShorthandObject;
```

>   dismissAction?: ShorthandValue<HTMLElement>;

Author: ```suggestion
  dismissAction?: ShorthandProps<HTMLElement>;
```

>   dismissAction?: ShorthandValue<AlertDismissActionProps>;

Author: ```suggestion
  dismissAction?: ShorthandValue<HTMLElement>;
```

# [Pills]()

## Background

PIlls should be used when representing an input, as a way to filter content, or to represent an attribute.

> ## Prior Art

Reviewer 4: Should this just be a link to Open UI research ?

Author: The research hasn't been merged yet 😔

Reviewer 4: just link the PR I guess :)

- [Evergreen](https://evergreen.segment.com/components/badge-and-pill/)
- [Lightning](https://www.lightningdesignsystem.com/components/pills/)

> - [OpenUI Research]()

Reviewer 5: Please don't forget to add the link here ;)

## Sample Code

```jsx
<Pills>
  <Pill>Item 1</Pill>
  <Pill>Item 2</Pill>
</Pills>
```

## Variants

- Shape: `circular`, `rounded`

> - Size: `smaller`, `small`, `medium`

Reviewer 5: Why only those variants? Does current design spec have only those?

Author: Yep, design spec only has that

> - Appearance: `Filled Grey (filled)`, `Filled White (inverted)`, `Outline`

Reviewer 5: Should we call-out specific color here?

## PROPS

```typescript
export type PillShape = 'rounded' | 'circular';

export type PillAppearence = 'filled' | 'inverted' | 'outline';

export type SizeValue = 'smaller' | 'small' | 'medium';

export interface PillProps extends ComponentProps, React.HTMLAttributes<HTMLElement> {
  /**
   * A Pill can be sized.
   */
  size?: SizeValue;

  /**
   * A Pill can be circular or rounded
   */
  shape?: PillShape;

  /**
   * A Pill can be filled, inverted or outline
   */
  appearance?: PillVariant;

  /**
   * Media slot
   */
  media?: ShorthandProps<HTMLElement>;

  /**
   * Details that will appear as second line
   */
  details?: ShorthandProps<HTMLElement>;

  /** A Pill can be dismissible. */
  dismissible?: boolean;

  /** A Pill can be disabled. */
  disabled?: boolean;

  /** A Pill can be selected. */
  selected?: boolean;

  /** A Pill can be draggable  */

>   draggable?: boolean;

Reviewer 4: You don't seem to mention draggable behaviour later on

  /**
   * A button shorthand for the dismiss action slot. To use this slot the pill should be
   * dismissible.
   */

>   dismissAction?: ShorthandValue<AlertDismissActionProps>;

Reviewer 5: Please are those `AlertDismissActionProps` from Fluent UI N*?

}
```

## Structure

- _**Public**_

```jsx
<Pills>
  <Pill />
</Pills>
```

- _**Internal**_

```jsx
<ElementType >
  <li>
    <span>
      {children}
    </span>
  <li>
</ElementType>
```

- _**DOM**_

```html

> <ul aria-label>

Reviewer 4: for markup flexibility, should we default to `div` with appropriate roles ?

Author: `<div role="list"`?

Reviewer 4: Yeah, although you sould check with our a11y champs if that role is even necessary

> <ul aria-label>

Author: ```suggestion
<div role="list" aria-label>
```

>   <li role="presentation">

Reviewer 4: I've never understood the value of having these `li` wrappers with `role ="presentation"`, it's purely to make sure that the usage with `ul` is correct. 

In this case doesn't it make more sense to use divs ?

Author: Well, I guess just to be semantically correct.

Reviewer 4: but why not just use `<li role="button">` ? 

presentation just kills the default list role anyway

    <span role="button">
      ...
    </span>
  </li>

> </ul>

Author: ```suggestion
</div>
```

```

## Migration

- _Migration from v8_

...

- _Migration from v0_

`Dropdown` component should use `Pills` over `DropdownSelectedItem`.

### Pill with dismiss

#### Accessibility

- Each pill is of role="button"
- Tab key to set focus on the Pill

> - Tab key moves focus to next item

Reviewer 4: Here do you mean that the tab key focuses on the next focusable item after the `Pills` group ?

Author: I mean, focus to next `Pill`

> - Tab key moves focus to next item

Author: ```suggestion
- Tab key moves focus to next `Pill`
```

> - Esc key remove focus

Reviewer 4: Removes focus and then focuses on what ?

Author: Actually, `Esc` should dismiss the `Pill`

> - Esc key remove focus

Author: ```suggestion
- Esc key dismiss the `Pill`
```

#### Keyboard Navigation

| key                | state    | title                                                              |
| ------------------ | -------- | ------------------------------------------------------------------ |
| left/right         | OnFocus  | Role= “button”, aria-describedby=”Press enter or delete to remove” |
| Enter/Delete       | OnDelete | The pill is removed since there is only one primary action         |
| Backspace/Ctrl + X | OnDelete | Backspace/Ctrl + X                                                 |

### Pill with popup

#### Accessibility

> - Each pill is of role="button"

Reviewer 4: if all pills are of `role="button"` we should state that earlier instead of requiring the reader to check each variation

> - Popup opens after 500ms of mouse hover state

Reviewer 4: Shuold we make the delay configurable ?

> - Popup opens after 500ms of mouse hover state

Reviewer 5: Sorry, I'm missing something - how pill and popup are connected? Each pill has a popup? What's the contents of the popup?

Author: It can happen as per Design Spec ( each pill having a popup )

> - Popup opens after 500ms of mouse hover state

Author: ```suggestion
- Popup opens after Xms of mouse hover state
```

- Clicking in the Pill open popup and clicking outside should dimiss it

#### Keyboard Navigation

| key                | state    | title                                                              |
| ------------------ | -------- | ------------------------------------------------------------------ |
| Esc                | OnFocus  | Dismis popup                                                       |
| Enter              | OnFocus  | Open Popup                                                         |
| Backspace/Ctrl + X | OnDelete | Backspace/Ctrl + X                                                 |

> | left/right         | OnFocus  | Role= “button”, aria-describedby=”Press enter or delete to remove” |

Reviewer 4: confused by `aria-describedby`, shouldn't it take an id of a DOM element ?

Reviewer 5: Yes it should https://www.w3.org/TR/wai-aria/#aria-describedby ;)

Author: Yeah, it should.

> | left/right         | OnFocus  | Role= “button”, aria-describedby=”Press enter or delete to remove” |

Author: ```suggestion
| left/right         | OnFocus  | Role= “button”, aria-label=”Press enter or delete to remove” |
```

### Pill with right click (invoking context menu)

#### Accessibility

- Each pill is of role="button"

> - Right click anywhere on card surface invokes content menu. Context menu appears close to cursor location on card

Reviewer 5: ```suggestion
- Right click anywhere on pills surface invokes content menu. Context menu appears close to cursor location on pill
```
Am I getting this right? Or what card are you referencing here?

- Clicking or tapping anywhere outside of menu to dismiss

#### Keyboard Navigation

| key       | state    | title                          |
| --------- | -------- | ------------------------------ |
| Esc       | OnFocus  | Dismis context menu            |
| Shift+F10 | OnFocus  | Open context menu              |

> | up/down   | Expanded | Navigates context menu options |

Reviewer 4: We should probably remove this line and state that once the menu is opened we follow menu behaviour, I think this should be exactly the standard menu used here

Author: Alright, I added just as reaffirmation

> | up/down   | Expanded | Navigates context menu options |

Author: ```suggestion
```

### Pill with toggle

#### Accessibility

- Each pill is of role="button"

> - Click toggles Pill's selected state

Reviewer 4: Do we need different aria-attributes or roles for this ? 

i.e. aria-selected/role="option" or if we stick with button then aria-pressed ?

Reviewer 4: DOM out put for each of the variations would be nice if they differ from the default

Author: I would go with `role="option"` I set as `button` as per design spec but don't think it's the best
